src/components/controller.cpp

```diff
@@ -386,36 +390,195 @@ void controller::process_eventqueue() {
+            // Forward to shell if the module cannot accept the action to not break existing behavior.
+            return false;
+          }
+          // Only deliver to the first matching module.
+          return true;
+        }
+      }
+    }
+  }
+
+  /*
+   * If we couldn't find any matching legacy action, we return false and let
+   * the command be forwarded to the shell
+   */
+  return false;
+}
+
 /**
  * Process stored input data
  */
 void controller::process_inputdata() {
-  if (!m_inputdata.empty()) {
-    string cmd = m_inputdata;
-    m_inputdata.clear();
+  if (m_inputdata.empty()) {
+    return;
+  }
+
+  const string cmd = m_inputdata;
+  m_inputdata.clear();
 
+  m_log.trace("controller: Processing inputdata: %s", cmd);
+
+  // Every command that starts with '#' is considered an action string.
+  if (cmd.front() == '#') {
+    string handler_name;
+    string action;
+    string data;
+
+    try {
+      auto res = actions_util::parse_action_string(cmd);
+
+      handler_name = std::get<0>(res);
+      action = std::get<1>(res);
+      data = std::get<2>(res);
+    } catch (runtime_error& e) {
+      m_log.err("Invalid action string (reason: %s)", e.what());
+      return;
+    }
+
+    m_log.info("Forwarding data to input handlers (name: '%s', action: '%s', data: '%s') ", handler_name, action, data);
+
+    int num_delivered = 0;
+
+    // Forwards the action to all input handlers that match the name
     for (auto&& handler : m_inputhandlers) {
-      if (handler->input(string{cmd})) {
-        return;
+      if (handler->input_handler_name() == handler_name) {
+        if (!handler->input(std::forward<string>(action), std::forward<string>(data))) {
+          m_log.err("The '%s' module does not support the '%s' action.", handler_name, action);
+        }
+
+        num_delivered++;
       }
     }
 
-    try {
-      m_log.info("Uncaught input event, forwarding to shell... (input: %s)", cmd);
+    if (num_delivered == 0) {
+      m_log.err("There exists no input handler with name '%s' (input: %s)", handler_name, cmd);
+    } else {
+      m_log.info("Delivered input to %d input handler%s", num_delivered, num_delivered > 1 ? "s" : "");
+    }
+    return;
+  }
 
-      if (m_command) {
-        m_log.warn("Terminating previous shell command");
-        m_command->terminate();
-      }
+  if (this->try_forward_legacy_action(cmd)) {
+    return;
+  }
 
-      m_log.info("Executing shell command: %s", cmd);
-      m_command = command_util::make_command<output_policy::IGNORED>(move(cmd));
-      m_command->exec();
-      m_command.reset();
-      process_update(true);
-    } catch (const application_error& err) {
-      m_log.err("controller: Error while forwarding input to shell -> %s", err.what());
+  try {
+    // Run input as command if it's not an input for a module
+    m_log.info("Forwarding command to shell... (input: %s)", cmd);
+
+    if (m_command) {
```

Comment:
Useless code no ? Since we are always resetting the pointer (L.578)

Reply:
That code was already there. But you are right. I guess the original intent was to let the command run while polybar does other things, but since we always wait for completion, that isn't the case.